cmd/clusterctl/client/cluster/objectgraph.go

```diff
@@ -309,6 +311,26 @@ func (o *objectGraph) Discovery(namespace string) error {
 			return err
 		}
 
+		// if we are discovering Secrets, also secrets from the providers namespace should be included.
+		if discoveryType.typeMeta.GetObjectKind().GroupVersionKind().GroupKind() == corev1.SchemeGroupVersion.WithKind("SecretList").GroupKind() {
+			providers, err := o.providerInventory.List()
+			if err != nil {
+				return err
+			}
+			for _, p := range providers.Items {
+				if p.Type == string(clusterctlv1.InfrastructureProviderType) {
+					providerNamespaceSelector := []client.ListOption{client.InNamespace(p.Namespace)}
```

Comment:
Perfect for the docs, can we make sure to also enforce this in code?

Reply:
This should be implemented on the provider side, might be @randomvariable has some ideas about how to enforce it..

Reply:
- default credentials mounted in the pods are within the provider namespace
- the `identityRef` shouldn't have a namespace, so that's also good (especially for providers who reference a secret directly)
- for providers referencing a CR or a cluster-scoped `identityRef` we'll need validation or a webhook the provider side if we need to look for which namespace the provider is installed